index.bs

```diff
@@ -2219,6 +2256,89 @@ To <dfn>parse trigger specs</dfn> given a [=map=] |map|, a [=moment=]
+    and |map|["`max_num_event_states`"] does not [=map/exist=], return an error.
+1. Let |maxNumEventStates| be the result of running
+    [=parse an optional 32-bit unsigned integer=] with |value|,
+    "`max_num_event_states`", and 3.
+1. If |maxNumEventStates| is an error or |maxNumEventStates| is 0, return an error.
+1. If |sourceType| is "<code>[=source type/event=]</code>" and |states| > |maxNumEventStates|, return an error.
+1. Return |maxNumEventStates|
+
+To <dfn>parse attribution scopes</dfn> from a [=map=] |map|, a possibly null 32-bit unsigned integer |attributionScopeLimit|:
+1. Let |result| be an [=ordered set=].
+1. If |map|["`attribution_scopes`"] [=map/exists|exist=]:
+    1. Let |values| be |map|["`attribution_scopes`"].
+    1. If |values| is not a [=list=], return an error.
+    1. [=list/iterate|For each=] |value| of |values|:
+        1. If |value| is not a string, return an error.
+        1. Let |attributionScope| be the result of applying the
+            <a spec="html">rules for parsing integers</a> to |value|.
+        1. If |attributionScope| cannot be represented by a 64-bit signed integer, return an error.
+        1. [=set/Append=] |attributionScope| to |result|.
+1. If |result| is not empty:
+    1. If |attributionScopeLimit| is null, return an error
+1. If |result|'s size > |attributionScopeLimit|, return an error.
+1. Return |result|.
+
+To <dfn>remove sources with different max num view states if necessary</dfn> given an [=attribution source=] |pendingSource|:
+1. If |pendingSource|'s [=attribution source/max num view states] is null, return.
+1. Let |sourcesToRemove| be a new [=list/is empty|empty=] [=list=].
+1. Let |cache| be the user agent's [=attribution source cache=].
+1. [=set/iterate|For each=] |source| of |cache|:
+    1. Let |commonDestinations| be [=set/intersection=] of |source|'s [=attribution source/attribution destinations=] and |pendingSource|'s [=attribution source/attribution destinations=].
+    1. if |commonDestinations| is empty, [=iteration/continue=].
+    1. If |source|'s [=attribution source/source origin=] and |pendingSource|'s [=attribution source/source origin=] are not [=same origin=], [=iteration/continue=].
+    1. If |source|'s =[attribution source/max num view states] is null, [=iteration/continue=].
+    1. if |source|'s =[attribution source/max num view states] equals |pendingSource|'s [=attribution source/max num view states], [=iteration/continue=].
+    1. [=list/Append=] |source| to |sourcesToRemove|.
+1. Remove |sourcesToRemove| from [=attribution source cache=].
+
+To <dfn>remove sources with smaller attribution scope limit from cache if necessary</dfn> given an [=attribution source=] |pendingSource|
+1. If |pendingSource|'s [=attribution source/attribution scope limit=] is null, return.
+1. Let |sourcesToRemove| be a new [=list/is empty|empty=] [=list=].
+1. Let |cache| be the user agent's [=attribution source cache=].
+1. [=set/iterate|For each=] |source| of |cache|:
```

Comment:
The body of this loop seems very similar to the one in `remove sources with different max num view states if necessary`. Is there a common sub-algorithm we can extract?

Reply:
Which part needs a lambda?

Reply:
Added a shared sub algorithm `find sources with common destinations and reporting origin`
(I was thinking to pass a lambda to the shared method, but realized it's not necessary)